rewrite intro to materialize (#335)
authzed/docs: pages/authzed/concepts/authzed-materialize.mdx

</Callout>

AuthZed Materialize is inspired by the Leopard index component described in the [Zanzibar paper](https://zanzibar.tech/2IoYDUFMAE:0:T).
Much like the concept of a materialized view in relational databases, AuthZed Materialize is a service that computes how permissions change after relationships
are written, when those relationships affect a subject's membership in a permission set or a set’s permission on a specific resource.
These permissions are then streamed to your client, so that you can store them as a denormalized table, and then do operations like searching, sorting, and filtering much more efficiently.

AuthZed Materialize allows you to:
- speed up `CheckPermission` and `CheckBulkPermissions`
- speed up `LookupResources` and `LookupSubjects`, especially when there is a large number of resources
- build authorization-aware UIs, e.g. by providing a filtered and/or sorted list of more than several thousand authorized objects
- perform ACL filtering in other secondary indexes, like a search index (e.g. ElasticSearch)

[Dedicated]: ../guides/picking-a-product#dedicated

## Limitations

If there are:

- [Caveats]
- [Wildcard] subject types
- [.all]

> on the path of permissions computed by Materialize, it will error out. However, your schema can still include them.

Author: To be confirmed: does it error out? It is just my assumption

Reviewer: correct, so long it's not in the reachability path of a watched permission, Materialize will not complain. If it does, the hydrator will fail

Author: can I mention "hydrator" here or is that too much information?

Reviewer: Let's skip if for simplicity

import { Callout } from 'nextra/components'

# AuthZed Materialize

<Callout type="info">
  AuthZed Materialize is available to users of AuthZed [Dedicated] as part of an early access program.
  Don't hesitate to get in touch with your AuthZed account team if you would like to participate.
</Callout>

> AuthZed Materialize is inspired by the Leopard index component described in the [Zanzibar paper](https://zanzibar.tech/2IoYDUFMAE:0:T).

Reviewer 2: s/is inspired by/takes inspiration from

> Much like the concept of a materialized view in relational databases, AuthZed Materialize is a service that computes how permissions change after relationships

Reviewer 2: ```suggestion
Much like the concept of a materialized view in relational databases, AuthZed Materialize precomputes specified permissions defined in your SpiceDB schema. 
```

are written, when those relationships affect a subject's membership in a permission set or a set’s permission on a specific resource.
These precomputed permissions can then be used either to provide faster checks and lookups through Accelerated Queries, or streamed to your own application database to do operations like searching, sorting, and filtering much more efficiently.

In summary, AuthZed Materialize allows you to:

- speed up `CheckPermission` and `CheckBulkPermissions`

> - speed up `LookupResources` and `LookupSubjects`, especially when there is a large number of resources

Reviewer 2: ```suggestion
- Speed up `LookupResources` and `LookupSubjects`, especially when there is a large number of resources
```

> - build authorization-aware UIs, e.g. by providing a filtered and/or sorted list of more than several thousand authorized objects

Reviewer 2: ```suggestion
- Build authorization-aware UIs, e.g. by providing a filtered and/or sorted list of more than several thousand authorized objects
```

> - perform ACL filtering in other secondary indexes, like a search index (e.g. ElasticSearch)

Reviewer 2: ```suggestion
- Perform ACL filtering in other secondary indexes, like a search index (e.g. ElasticSearch)
```

import { Callout } from 'nextra/components'

# AuthZed Materialize

<Callout type="info">
  AuthZed Materialize is available to users of AuthZed [Dedicated] as part of an early access program.
  Don't hesitate to get in touch with your AuthZed account team if you would like to participate.
</Callout>

AuthZed Materialize is inspired by the Leopard index component described in the [Zanzibar paper](https://zanzibar.tech/2IoYDUFMAE:0:T).
Much like the concept of a materialized view in relational databases, AuthZed Materialize is a service that computes how permissions change after relationships
are written, when those relationships affect a subject's membership in a permission set or a set’s permission on a specific resource.

> These permissions are then streamed to your client, so that you can store them as a denormalized table, and then do operations like searching, sorting, and filtering much more efficiently.

Reviewer 3: This is one way to use it; the other is "Accelerated Queries," which I think is what we want to default users to, since it doesn't require additional work on their side.
```suggestion
These precomputed permissions can then either be used to provide faster checks and lookups through Accelerated Queries or streamed to your own application database to do operations like searching, sorting, and filtering much more efficiently.
```

AuthZed Materialize allows you to:

- speed up `CheckPermission` and `CheckBulkPermissions`
- speed up `LookupResources` and `LookupSubjects`, especially when there is a large number of resources
- build authorization-aware UIs, e.g. by providing a filtered and/or sorted list of more than several thousand authorized objects
- perform ACL filtering in other secondary indexes, like a search index (e.g. ElasticSearch)

[Dedicated]: ../guides/picking-a-product#dedicated

> ## Limitations

Author: Added, let me know if you like the wording
